### PR TITLE
DRA7xx fixups

### PR DESCRIPTION
**Pull request by glneo**
A couple changes needed to restore OPTEE functionality with our latest U-Boot/Linux and some small fixups.

### CODE REVIEW

**Comment by igoropaniuk** [comment]
Hi @glneo ,

There is a lot of warnings from checkpatch.pl about lines over 80 characters (check travis-ci build log) in your changes
`WARNING: line over 80 characters`

BTW, you can use our checkpatch.sh wrapper for checkpatch.pl for convenient checking of working/staging area before committing.

File: core/arch/arm/plat-ti/console.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <platform_config.h>
+#include <drivers/serial8250_uart.h>
+#include <mm/core_memprot.h>
+#include <console.h>
```
Comment (igoropaniuk) [inline core/arch/arm/plat-ti/console.c]: Please, sort in alphabetical order. Thanks!

File: core/arch/arm/plat-ti/main.c [hunk 2]
```diff
@@ -130,12 +93,13 @@ struct plat_nsec_ctx {
 
 void init_sec_mon(unsigned long nsec_entry)
 {
-	struct plat_nsec_ctx *plat_ctx = (struct plat_nsec_ctx *)nsec_entry;
+	struct plat_nsec_ctx *plat_ctx = (struct plat_nsec_ctx *)
+		(phys_to_virt(nsec_entry, MEM_AREA_IO_SEC));
```
Comment (igoropaniuk) [inline core/arch/arm/plat-ti/main.c]:
```
if (!plat_ctx)
	panic();
```

File: core/arch/arm/plat-ti/api_monitor_index.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES// LOSS OF USE, DATA, OR PROFITS// OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __API_MONITOR_INDEX_H__
```
Comment (igoropaniuk) [inline core/arch/arm/plat-ti/api_monitor_index.h]: Why wrap with __ __?

File: core/arch/arm/plat-ti/console.c (hunk 1 above)
Comment (glneo) [inline core/arch/arm/plat-ti/console.c]: Will do.

File: core/arch/arm/plat-ti/main.c (hunk 2 above)
Comment (glneo) [inline core/arch/arm/plat-ti/main.c]: Makes sense, will add.

File: core/arch/arm/plat-ti/api_monitor_index.h (hunk 3 above)
Comment (glneo) [inline core/arch/arm/plat-ti/api_monitor_index.h]: This file is pulled from the DRA7 platform firmware, we would like to keep it as close to the original as possible. I can remove the __, but changing the names of variables to make them shorter (to help fix the +80 char lines) will reduce the cross-referenceability and moving the defines all to new lines really reduces readability.

**Comment by jforissier** [comment]
On Mon, Nov 14, 2016 at 9:22 PM, Andrew F. Davis <notifications@github.com>
wrote:

> *@glneo* commented on this pull request.
> ------------------------------
>
> In core/arch/arm/plat-ti/api_monitor_index.h
> <https://github.com/OP-TEE/optee_os/pull/1180>:
>
> > + * and/or other materials provided with the distribution.
> + *
> + * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
> + * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
> + * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
> + * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
> + * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
> + * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
> + * SUBSTITUTE GOODS OR SERVICES// LOSS OF USE, DATA, OR PROFITS// OR BUSINESS
> + * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
> + * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
> + * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
> + * POSSIBILITY OF SUCH DAMAGE.
> + */
> +
> +#ifndef __API_MONITOR_INDEX_H__
>
> This file is pulled from the DRA7 platform firmware, we would like to keep
> it as close to the original as possible. I can remove the __, but changing
> the names of variables to make them shorter (to help fix the +80 char
> lines) will reduce the cross-referenceability and moving the defines all to
> new lines really reduces readability.
>
Shared files are a valid case for adding an exception to our checkpath
procedure, see
https://github.com/OP-TEE/optee_os/blob/master/scripts/checkpatch_inc.sh#L5.
I have no strong opinion on this, so if you feel more comfortable with
keeping the original file unchanged, then do it but in this case please
update checkpatch_inc.sh.

File: scripts/checkpatch_inc.sh [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-CHECKPATCH=checkpatch.pl
+CHECKPATCH="${CHECKPATCH:-checkpatch.pl}"
```
Comment (igoropaniuk) [inline scripts/checkpatch_inc.sh]: Why not just
`export PATH=/path/where/checkpatch.pl/stored/:$PATH`

File: core/arch/arm/sm/sm_a32.S [hunk 5]
```diff
@@ -139,9 +139,23 @@ UNWIND(	.cantunwind)
 	orr	r0, r0, #(SCR_NS | SCR_FIQ) /* Set NS and FIQ bit in SCR */
 	write_scr r0
 
+	dsb
+	isb
+
 	b	.smc_exit
 
 .smc_ret_to_sec:
+	/*
```
Comment (jenswi-linaro) [inline core/arch/arm/sm/sm_a32.S]: This is just overhead for platforms not needing it.
It also makes it harder to make future changes to the secure monitor as the TI sm_platform_handler depends on internal stuff of the secure monitor.

File: scripts/checkpatch_inc.sh (hunk 4 above)
Comment (glneo) [inline scripts/checkpatch_inc.sh]: That will still work, I just like being able to easily change tools from the command line if I need, without messing with my PATH.

File: core/arch/arm/sm/sm_a32.S [hunk 6]
```diff
@@ -139,9 +139,23 @@ UNWIND(	.cantunwind)
 	orr	r0, r0, #(SCR_NS | SCR_FIQ) /* Set NS and FIQ bit in SCR */
 	write_scr r0
 
+	dsb
```
Comment (glneo) [inline core/arch/arm/sm/sm_a32.S]: Hmmm, now that I think about it, I'm not sure, I'll remove them and see if they had a purpose.

File: core/arch/arm/sm/sm_a32.S (hunk 5 above)
Comment (glneo) [inline core/arch/arm/sm/sm_a32.S]: I'll rebase on #1184 and clean this up a bit.

Comment (glneo) [inline core/arch/arm/sm/sm_a32.S]: Now that I look into it, I think I will drop this patch for now, I'll push it sometime after #1184 gets merged.

**Comment by glneo** [comment]
It looks like the travis-ci check fail is a false positive checkpatch.pl warning. Are there any other comments that need addressing for this series?

**Comment by igoropaniuk** [comment]
Hi @glneo,
It's failing because of this a bit tricky warning :) 
> WARNING: A patch subject line should describe the change not the tool that found it

See [check] in checkpatch.p(https://github.com/torvalds/linux/blob/master/scripts/checkpatch.pl#L2483)
Seems that "checkpatch" shouldn't be mentioned in the commit subject

**Comment by jforissier** [comment]
@igoropaniuk IMO this warning should be ignored, because the commit subject does indeed describe the change properly. It would not make much sense to remove "checkpatch" from the sentence. Would you prefer: "4ll0w 0v3rwriting t3h ch3ckp4tch.pl l0c4ti0n" maybe? ;-)

File: core/arch/arm/plat-ti/main.c [hunk 7]
```diff
@@ -130,12 +93,16 @@ struct plat_nsec_ctx {
...
 	struct sm_nsec_ctx *nsec_ctx;
 
+	plat_ctx = phys_to_virt(nsec_entry, MEM_AREA_IO_SEC);
+	if (!plat_ctx)
+		panic();
+
 	/* Invalidate cache to fetch data from external memory */
-	cache_maintenance_l1(DCACHE_AREA_INVALIDATE, (void *)nsec_entry,
-		sizeof(struct plat_nsec_ctx));
+	cache_maintenance_l1(DCACHE_AREA_INVALIDATE,
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-ti/main.c]: nit: The cast shouldn't be needed.

**Comment by igoropaniuk** [comment]
@jforissier actually, that's what I did when I was adding new commits related to checkpatch.sh in optee_os/scripts :)

@glneo 
`Reviewed-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

File: core/arch/arm/plat-ti/console.c [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+
+register_phys_mem(MEM_AREA_IO_NSEC,
+		  CONSOLE_UART_BASE,
+		  SERIAL8250_UART_REG_SIZE);
+
+static vaddr_t console_base(void)
+{
+	static void *va;
```
Comment (etienne-lms) [inline core/arch/arm/plat-ti/console.c]: `static void *va __early_bss;`

File: core/arch/arm/plat-ti/main.c [hunk 9]
```diff
@@ -130,12 +93,16 @@ struct plat_nsec_ctx {
 
 void init_sec_mon(unsigned long nsec_entry)
 {
-	struct plat_nsec_ctx *plat_ctx = (struct plat_nsec_ctx *)nsec_entry;
+	struct plat_nsec_ctx *plat_ctx;
 	struct sm_nsec_ctx *nsec_ctx;
 
+	plat_ctx = phys_to_virt(nsec_entry, MEM_AREA_IO_SEC);
```
Comment (etienne-lms) [inline core/arch/arm/plat-ti/main.c]: nsec_entry is mapped secure ? was it even declared in optee core mapping ?

also: if it was mapped as `MEM_AREA_IO_SEC`, then it is mapped uncached. Do you really need to invalidate the cache as done below ?

Comment (jenswi-linaro) [inline core/arch/arm/plat-ti/main.c]: @etienne-lms, in plat-ti nsec_entry is not the same thing as in other platforms.

Comment (etienne-lms) [inline core/arch/arm/plat-ti/main.c]: right. thanks.

i stll wonder if cache invalidation below as required.

File: core/arch/arm/plat-ti/main.c (hunk 7 above)
Comment (glneo) [inline core/arch/arm/plat-ti/main.c]: Removed.

File: core/arch/arm/plat-ti/console.c (hunk 8 above)
Comment (glneo) [inline core/arch/arm/plat-ti/console.c]: Added.

File: core/arch/arm/plat-ti/main.c (hunk 9 above)
Comment (glneo) [inline core/arch/arm/plat-ti/main.c]: Thinking more on the cache invalidation, I don't think it is strictly needed. As you said this area will be mapped un-cached as it is marked MEM_AREA_IO_SEC, but this is area is not IO area, it is just marked this way. A better fix would be to mark this area MEM_AREA_RAM_SEC, which it is, then the invalidation may be needed (this location is written by the platforms default TEE so I'm not sure about its cache state before the hand-off).

Comment (etienne-lms) [inline core/arch/arm/plat-ti/main.c]: @glneo: using MEM_AREA_IO_SEC or MEM_AREA_RAM_SEC is just a... trick to define physical address ranges as cached or uncached (both secure of course), whatever the physical HW lies below. So it's fine using MEM_AREA_IO_SEC for uncached RAM.

- If your expected data were written through the cache, then you should map it cached and no need to flush or invalidate it previously.
- If the data were written uncached, then map it uncached and no need to flush or invalidate.
- If your not unsure... it's a tricky situation. Mapping cached and invalidating cache may discard content that were cached. 

I think you should have a look how the earlier boot stage handled these data. Or maybe you should flush the overall cache at optee entrypoint. Actually the earlier boot stage should have flushed its cache for data accessed by its next boot stage before entering it (here optee).

**Comment by jforissier** [comment]
@glneo would you mind rebasing on top of optee_os/master please?

**Comment by jforissier** [comment]
Rebased on top of master and merged. Thanks @glneo.

**Pull request closed without merging**